subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/RoutingVirtualFileSystem.java

```diff
@@ -0,0 +1,102 @@
+    private final WellKnownFileLocations gradleUserHomeFileLocations;
+    private final VirtualFileSystem gradleUserHomeVirtualFileSystem;
+    private final VirtualFileSystem buildScopedVirtualFileSystem;
+
+    public RoutingVirtualFileSystem(
+        WellKnownFileLocations gradleUserHomeLocations,
+        VirtualFileSystem gradleUserHomeVirtualFileSystem,
+        VirtualFileSystem buildScopedVirtualFileSystem
+    ) {
+        this.gradleUserHomeFileLocations = gradleUserHomeLocations;
+        this.gradleUserHomeVirtualFileSystem = gradleUserHomeVirtualFileSystem;
+        this.buildScopedVirtualFileSystem = buildScopedVirtualFileSystem;
+    }
+
+    @Override
+    public void read(String location, FileSystemSnapshotVisitor visitor) {
+        getVirtualFileSystemFor(location).read(location, visitor);
+    }
+
+    @Override
+    public <T> Optional<T> readRegularFileContentHash(String location, Function<HashCode, T> visitor) {
+        return getVirtualFileSystemFor(location).readRegularFileContentHash(location, visitor);
+    }
+
+    @Override
+    public void read(String location, SnapshottingFilter filter, FileSystemSnapshotVisitor visitor) {
+        getVirtualFileSystemFor(location).read(location, filter, visitor);
+    }
+
+    @Override
+    public void update(Iterable<String> locations, Runnable action) {
+        if (Iterables.isEmpty(locations)) {
+            return;
+        }
+        Iterable<String> immutableLocations = Iterables.filter(locations, gradleUserHomeFileLocations::isImmutable);
+        int immutableLocationsSize = Iterables.size(immutableLocations);
+        if (immutableLocationsSize == 0) {
+            buildScopedVirtualFileSystem.update(locations, action);
+        } else if (immutableLocationsSize == Iterables.size(locations)) {
+            gradleUserHomeVirtualFileSystem.update(locations, action);
+        } else {
+            Iterable<String> mutableLocations = Iterables.filter(locations, location -> !gradleUserHomeFileLocations.isImmutable(location));
+            gradleUserHomeVirtualFileSystem.update(immutableLocations, action);
+            buildScopedVirtualFileSystem.update(mutableLocations, action);
+        }
+    }
+
+    @Override
+    public void invalidateAll() {
+        gradleUserHomeVirtualFileSystem.invalidateAll();
+        buildScopedVirtualFileSystem.invalidateAll();
+    }
+
+    @Override
+    public void updateWithKnownSnapshot(String location, FileSystemLocationSnapshot snapshot) {
+        getVirtualFileSystemFor(location).updateWithKnownSnapshot(location, snapshot);
+    }
+
+    @Override
+    public FileSystemSnapshot snapshotWithBuilder(Consumer<FileSystemSnapshotBuilder> buildAction) {
+        return buildScopedVirtualFileSystem.snapshotWithBuilder(buildAction);
+    }
+
+    private VirtualFileSystem getVirtualFileSystemFor(String location) {
+        return gradleUserHomeFileLocations.isImmutable(location)
```

Comment:
Probably not for this PR to fix, but we have several terms distinguishing between the options here: `mutable` vs `immutable`, `well-known` vs. ???, `gradle user home` vs `build scoped`. I think we should try to reduce the number of terms. Not sure how exactly.

Reply:
Makes sense.